### PR TITLE
associative containers: added method to retrieve a value by key and the ability to discern containers with mapped types from simple keyed containers

### PR DESCRIPTION
**Pull request by alexmontAmazon**
added method to discern between associative containers with mapped
values and simple containers with only keys. Added GetValueByKey to
retrieve values from associative containers by key.

Signed-off-by: Alex Montgomery <alexmont@amazon.com>

## What does this PR do?
added method to discern between associative containers with mapped
values and simple containers with only keys. Added GetValueByKey to
retrieve values from associative containers by key.

## How was this PR tested?
locally with Generic DOM code

### CODE REVIEW

File: Code/Framework/AzCore/AzCore/Serialization/AZStdContainers.inl [hunk 1]
```diff
@@ -48,6 +48,16 @@ namespace AZStd
...
     class shared_ptr;
+
+    namespace AssociativeInternal
+    {
+        // SFINAE expression to determine whether an associative container is an actual map with a corresponding value for each key,
+        // or is simply a keyed container, like a set or an unordered_set
+        template<class T, class = void>
+        constexpr bool IsMapType_v = false;
+        template<class T>
+        constexpr bool IsMapType_v<T, enable_if_t<Internal::sfinae_trigger_v<typename T::mapped_type>>> = true;
```
Comment (cgalvan) [inline Code/Framework/AzCore/AzCore/Serialization/AZStdContainers.inl:59]: Nice!

File: Code/Framework/AzCore/AzCore/Serialization/SerializeContext.h [hunk 2]
```diff
@@ -783,6 +783,8 @@ namespace AZ
                 virtual void*   GetElementByKey(void* instance, const ClassElement* classElement, const void* key) = 0;
                 /// Populates element with key (for associative containers). Not used for serialization.
                 virtual void    SetElementKey(void* element, void* key) = 0;
+                /// Get the mapped value's address by its key. If there is no mapped value (like in a set<>) the value returned is the key itself
```
Comment (cgalvan) [inline Code/Framework/AzCore/AzCore/Serialization/SerializeContext.h:786]: Should this be `//!` ? I'm actually not sure what the difference is between `//!` and `///`

**Review by cgalvan — APPROVED** [review]

**Review by lumberyard-employee-dm — APPROVED** [review]

File: Code/Framework/AzCore/AzCore/Serialization/AZStdContainers.inl (hunk 1 above)
Comment (alexmontAmazon) [reply Code/Framework/AzCore/AzCore/Serialization/AZStdContainers.inl:59]: yep! Courtesy of our guru @lumberyard-employee-dm  👍

File: Code/Framework/AzCore/AzCore/Serialization/SerializeContext.h (hunk 2 above)
Comment (alexmontAmazon) [reply Code/Framework/AzCore/AzCore/Serialization/SerializeContext.h:786]: I think both work for doxygen. I was just trying to adhere to the current style of that file

**Comment by lumberyard-employee-dm** [comment]
There is a conflict with how the AssociativeInternal namespace was opened up.
```
[2022-08-19T16:48:26.952Z]   ../../../../../../Code/Framework/AzCore/.\AzCore/Serialization/AZStdContainers.inl:52:15: error: inline namespace reopened as a non-inline namespace [-Werror]

[2022-08-19T16:48:26.952Z]       namespace AssociativeInternal

[2022-08-19T16:48:26.952Z]                 ^

[2022-08-19T16:48:26.952Z]       inline 

[2022-08-19T16:48:26.952Z]   ../../../../../../Code/Framework/AzCore/.\AzCore/std/containers/node_handle.h:201:22: note: previous definition is here

[2022-08-19T16:48:26.952Z]       inline namespace AssociativeInternal

[2022-08-19T16:48:26.952Z]                        ^

[2022-08-19T16:48:26.952Z]   1 error generated.
```
You should open AssociativeInternal namespace with the `inline` keyword to the match the existing namespace from the node handle header

**Review by lumberyard-employee-dm — APPROVED** [review]

**Pull request merged**